src/Our.Umbraco.Vorto/Web/UI/App_Plugins/Vorto/js/vorto.js

```diff
@@ -318,77 +319,107 @@
-                    .then(function (languages) {
+            // Load the languages (this will trigger everything else to bind)
+            languagesPromise.then(function(languages) {
+                setLanguages(languages);
+            });
+        }
 
-                        $scope.languages = languages;
+        var setDataType = function(dataType) {
+            // Stash the config in scope for reuse
+            $scope.property.config = dataType.preValues;
 
-                        if (!$scope.model.value.values) {
-                            $scope.model.value.values = {};
-                        }
+            // Get the view path
+            $scope.property.viewPath = umbPropEditorHelper.getViewPath(dataType.view);
 
-                        _.each($scope.languages, function (language) {
-                            if (!$scope.model.value.values.hasOwnProperty(language.isoCode)) {
-                                $scope.model.value.values[language.isoCode] = $scope.model.value.values[language.isoCode];
-                            }
-                        });
+            // Get the property alias
+            var propAlias = $scope.model.propertyAlias || $scope.model.alias;
 
-                        $scope.currentLanguage = $scope.activeLanguage = _.find(languages, function (itm) {
-                            return itm.isDefault;
-                        });
+            // Get the content type alias
+            var contentTypeAlias = nodeContext.contentTypeAlias || nodeContext.alias;
 
-                        reSync();
+            // Work out what section we are in
+            var currentSection = getCurrentSection();
 
-                        validateProperty();
+            var dataTypeByAliasKey = 'getDataTypeByAlias_' + currentSection + '_' + contentTypeAlias + '_' + propAlias;
+            var dataTypeByAliasPromise = cacheService.get(dataTypeByAliasKey);
+            if (!dataTypeByAliasPromise) {
+                dataTypeByAliasPromise = vortoResources.getDataTypeByAlias(currentSection, contentTypeAlias, propAlias);
+                cacheService.set(dataTypeByAliasKey, dataTypeByAliasPromise);
+            }
 
-                        detectFilledInLanguages();
-                    });
+            // Get the current properties datatype
+            dataTypeByAliasPromise.then(function(dataType2) {
+                setDataType2(dataType2, currentSection);
             });
-        });
+        };
 
+        var dataTypeByIdKey = "getDataTypeById_" + $scope.model.config.dataType.guid;
+        var dataTypeByIdPromise = cacheService.get(dataTypeByIdKey);
+        if (!dataTypeByIdPromise) {
+            // Load the datatype
+            dataTypeByIdPromise = vortoResources.getDataTypeById($scope.model.config.dataType.guid);
+            cacheService.set(dataTypeByIdKey, dataTypeByIdPromise);
```

Comment:
I don't really like that we are caching promises. I feel we should be caching the values as I think that's what people would expect. Maybe just have an if and call `setDataType` twice, once directly if we have a cache value, and once via a promise chain if we don't yet have a cached value.

Reply:
I've had another look into this and it seems to me to be a pretty good approach. The first call would create the promise which is cached, then all subsequent calls simply just use the .then function to set the relevant item. The API is only called once. Trying to cache the result only would result in the call being made several times until the API call is complete, thus avoiding the point of the change. I don't want to make the API calls block execution so went for reusing the promise on each call, which means the API is called just once.